internal/sessions/session_store.go

```diff
@@ -0,0 +1,310 @@
+package sessions
+
+import (
+	"fmt"
+	"log/slog"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/authentication"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/config"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/gwerrors"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/models"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/utils"
+	"github.com/labstack/echo/v4"
+	"github.com/redis/go-redis/v9"
+)
+
+// SessionStore handles sessions for the login server and the revproxy server
+type SessionStore struct {
+	authenticator  authentication.Authenticator
+	cookieTemplate func() http.Cookie
+	sessionMaker   SessionMaker
+	sessionRepo    models.SessionRepository
+	tokenStore     models.TokenStoreInterface
+}
+
+// Middleware returns the session middleware which injects the current session in the request context
+func (sessions *SessionStore) Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			session, loadErr := sessions.Get(c)
+			if loadErr != nil && loadErr != gwerrors.ErrSessionNotFound && loadErr != gwerrors.ErrSessionExpired {
+				slog.Info(
+					"SESSION MIDDLEWARE",
+					"message",
+					"could not load session",
+					"error",
+					loadErr,
+					"requestID",
+					utils.GetRequestID(c),
+				)
+			}
+			c.Set(SessionCtxKey, session)
+			err := next(c)
+			saveErr := sessions.Save(c)
```

Comment:
Why should a request fail if the saving session operation fails? This doesn't seem sound to me. E.g. if I hit `/api/data/projects` why return a 5xx if the session save fails?

Note that the session is touched and saved on every request (it extends the lifetime according to the idle rules). If a transient failure happens once, it should not result in a 5xx.

If there is a complete failure on the redis side, then `logging in` will not work (because saving/loading sessions will not), so I don't think we will cause more damage here.

Saving is done at the end because handlers can modify sessions, which is how `login` and `callback` works. Saving before the handlers would just not work.

Reply:
Note that this is similar to how `gorilla` sessions work, as many other session stores from other languages and frameworks.